cmake: clang: Support host's clang for non-MCU x86 targets on Linux

ozhuraki opened the pull request:
Pick host's clang if ZEPHYR_TOOLCHAIN_VARIANT=llvm.

This has been tested on samples/net/sockets/echo native_posix application with clang-7 and upstream’s version.

Compiles, links and runs with and without optimizations.

Instructions:

```
# sudo apt-get install clang
# cd samples/net/sockets/echo
# export ZEPHYR_TOOLCHAIN_VARIANT=llvm
# mkdir build
# cd build
# cmake -D BOARD=native_posix ..
```

dbkinder reviewed (approved):
doc change LGTM...

andrewboie commented on cmake/compiler/host-clang/target.cmake:
```diff
@@ -0,0 +1,20 @@
+
+find_program(CMAKE_C_COMPILER   clang   )
+#find_program(CMAKE_CXX_COMPILER clang++ )    # Not used currently
+#find_program(CMAKE_OBJCOPY     objcopy )     # Not used currently
```
Are you sure about that? I know a bunch of x86 post build steps use objcopy.
I'd rather we not commit commented-out code.

ozhuraki replied:
@andrewboie 

I just hecked for native_posix, it’s used at CMake configure/generate step, but not at compile step.

Sure, let’s include it.

mped-oticon commented on arch/posix/CMakeLists.txt:
```diff
@@ -1,12 +1,17 @@
   -include ${ZEPHYR_BASE}/arch/posix/include/posix_cheats.h
   )
 
+if(NOT CMAKE_C_COMPILER_ID STREQUAL "Clang") # clang errors on -fno-freestanding
```
There are many toolchains that are not clang.
I'd suggest the inverse: If GNU, rather than if not Clang.

ozhuraki replied:
@mped-oticon 

Makes sense. Fixed, have a look.

mped-oticon commented on cmake/compiler/gcc/target_security_fortify.cmake:
This file is for gcc, but until now clang has been reusing gcc stuff. But now there is a difference, so now clang does need its own `target_security_fortify.cmake`.

mped-oticon commented on cmake/compiler/host-clang/target.cmake:
```diff
@@ -0,0 +1,20 @@
+
+# Load toolchain_cc-family macros
+# Significant overlap with freestanding gcc compiler so reuse it
+include(${ZEPHYR_BASE}/cmake/compiler/gcc/target_security_fortify.cmake)
```
This should be `${ZEPHYR_BASE}/cmake/compiler/${COMPILER}/target_security_fortify.cmake`

ozhuraki replied:
I just checked, clang doesn't understand fortify at all. Fixed, have a look.

SebastianBoe commented on arch/posix/CMakeLists.txt:
```diff
@@ -1,12 +1,17 @@
   )
 
+if(NOT CMAKE_C_COMPILER_ID STREQUAL "Clang") # clang errors on -fno-freestanding
+zephyr_compile_options(
```
We can and should avoid this if statement by doing zephyr_cc_options() instead of zephyr_compile_options.

SebastianBoe commented on cmake/compiler/gcc/target_security_fortify.cmake:
This isn't sane. We are doing "if clang" in a gcc file. Clang should instead not use gcc's target_security_fortify.cmake.

ozhuraki replied:
@SebastianBoe 

You are right. I checked, clang doesn't understand fortify at all.  I refined it, please take a look.

SebastianBoe commented:
Would it be possible/desirable to have clang and llvm support host toolchains instead of introducing host-clang and host-llvm?

mped-oticon reviewed (changes requested):
Looks fine, a few minor changes required though.

(Eventually, the toolchain abstraction work will also abstract the linker and binutils, so TOPT will be dealt with)

ozhuraki commented:
@SebastianBoe 
> Would it be possible/desirable to have clang and llvm support host toolchains instead of introducing host-clang and host-llvm?

1)	It would be nice to simplify this a bit in all senses (including the elimination of host-llvm and host-clang folders), but I am not sure, I understand you here:

-	Merge cmake/toolchain/llvm and cmake/toolchain/host-llvm, cmake/compiler/clang and cmake/compiler/host-clang?
-	Drop cmake/toolchain/host-llvm and cmake/compiler/host-clang, if feasible?
-	-DCMAKE_C_COMPILER=clang while configuring the app?
-	Something else?

2)	What cmake/toolchain/llvm and cmake/compiler/clang are currently used for, for cross targets?

ozhuraki commented:
@nashif @SebastianBoe @andrewboie @mped-oticon 

And a wish for the native_posix is to make it optionally possible to relax –m32, so any host library can be linked into as is.

SebastianBoe commented:
> Merge cmake/toolchain/llvm and cmake/toolchain/host-llvm, cmake/compiler/clang and cmake/compiler/host-clang?

I'm suggesting to merge them, or in other words, have this PR patch the original build scripts instead of creating new ones.

aescolar commented:
> And a wish for the native_posix is to make it optionally possible to relax –m32

Allowing people to override it would be ok.
But note that the assumption that both pointers and ints are 32 bits is pervasive, and they are cast blindly back and forth in too many places. So the default should be that we compile for 32 bits.

nashif commented on cmake/target_toolchain.cmake:
```diff
@@ -30,7 +30,7 @@ set(CMAKE_SYSTEM_VERSION ${PROJECT_VERSION})
 set(BUILD_SHARED_LIBS OFF)
 
-if(NOT (COMPILER STREQUAL "host-gcc"))
+if(NOT ((COMPILER STREQUAL "host-gcc") OR (COMPILER STREQUAL "host-clang")))
```
how is host-clang different from the current clang compiler definition we have?

ozhuraki replied:
@nashif 

> how is host-clang different from the current clang compiler definition we have?

Not too much. After the discussion with @SebastianBoe, I am in progress of merging this with the existing cmake/toolchain/llvm and cmake/compiler/clang.

However I am missing something subtle there and can’t get it to work that way yet.

The original need is to use libFuzzer to fuzz a few states in the experimental TCP compiled as a native_posix application.

ozhuraki commented:
@SebastianBoe, @nashif

Merged with the existing cmake/toolchain/llvm and cmake/compiler/clang, take a look.

ozhuraki commented:
@SebastianBoe, @nashif

Sorry, I pushed a wrong stuff, ignore please.

ozhuraki commented:
@SebastianBoe, @nashif

This is the correct one, please take a look.

SebastianBoe commented on CMakeLists.txt:
I believe that by-design toolchains that don't support this should declare NO-OP functions.

Correct me if I'm wrong @mped-oticon

SebastianBoe commented on cmake/compiler/clang/generic.cmake:
```diff
@@ -1 +1,5 @@
-find_program(CMAKE_C_COMPILER clang PATH ${TOOLCHAIN_HOME} NO_DEFAULT_PATH)
+if(NOT "${ARCH}" STREQUAL "posix")
```
Remove the NOT

SebastianBoe commented on cmake/compiler/clang/generic.cmake:
```diff
@@ -1 +1,5 @@
+  find_program(CMAKE_C_COMPILER clang PATH ${TOOLCHAIN_HOME} NO_DEFAULT_PATH)
+else()
+  find_program(CMAKE_C_COMPILER clang)
+endif()
```
```
if(DEFINED TOOLCHAIN_HOME)
  set(find_program_clang_args PATH ${TOOLCHAIN_HOME} NO_DEFAULT_PATH)
endif()

find_program(CMAKE_C_COMPILER clang ${find_program_clang_args})
```

We want to reduce the code duplication here.

SebastianBoe replied:
The same technique should be used in target.cmake.

SebastianBoe reviewed (changes requested):
The build scripts need to be cleaned up.

See comments.

ozhuraki commented:
@SebastianBoe 
 
> The build scripts need to be cleaned up.

All these make sense, thanks, will redo & resend.

ozhuraki committed:
cmake: clang: Make clang pick -T option on linking

When linking, clang doesn't pick -T for some reason and complains,
while -Wl,-T works for both, gcc and clang.

Signed-off-by: Oleg Zhurakivskyy <oleg.zhurakivskyy@intel.com>

ozhuraki committed:
posix: cmake: Abstract freestanding options

clang has problems compiling the native_posix with -fno-freestanding.

Signed-off-by: Oleg Zhurakivskyy <oleg.zhurakivskyy@intel.com>

ozhuraki committed:
doc/getting_started: mention ZEPHYR_TOOLCHAIN_VARIANT=llvm

Mention that host installed clang can be used for native_posix.

Signed-off-by: Oleg Zhurakivskyy <oleg.zhurakivskyy@intel.com>

ozhuraki committed:
cmake: clang: Allow host's clang to be used for native_posix

Allow host installed clang to be used for native_posix when
ZEPHYR_TOOLCHAIN_VARIANT=llvm.

Signed-off-by: Oleg Zhurakivskyy <oleg.zhurakivskyy@intel.com>

ozhuraki committed:
toolchain: Do not pollute TOOLCHAIN_HOME

Use it's own variable HOST_TOOLS_HOME for host tools and don't
unconditionally set TOOLCHAIN_HOME, preventing the detection of
llvm/clang host toolchain.

Signed-off-by: Oleg Zhurakivskyy <oleg.zhurakivskyy@intel.com>

ozhuraki committed:
cmake: llvm: Make the setup for host installed clang easier

The logic is practically intact and is the following:

1. Use any host installed llvm/clang in the path in case
   ZEPHYR_TOOLCHAIN_VARIANT=llvm is requested alone.
2. This can be further restricted with TOOLCHAIN_HOME.
3. And can be further overridden with CLANG_ROOT_DIR,
   like previously.

So, only the unconditional restriction to /usr is lifted.

Together with fixing the unconditional set of TOOLCHAIN_HOME
by host tools for non-toolchain needs, this makes the logic
more flexible.

Now, after the logic is controllable by TOOLCHAIN_HOME, 3)
might be an extra, but is left intact for backward compatibility.

Signed-off-by: Oleg Zhurakivskyy <oleg.zhurakivskyy@intel.com>

ozhuraki committed:
cmake: clang: Detect host installed clang

In case TOOLCHAIN_HOME isn't explicitly reuqested,
(or indirectly forced with CLANG_ROOT_DIR), detect
any host installed clang in the path.

Signed-off-by: Oleg Zhurakivskyy <oleg.zhurakivskyy@intel.com>

ozhuraki committed:
cmake: clang: Fix comment

Fix comment, llvm -> clang.

Signed-off-by: Oleg Zhurakivskyy <oleg.zhurakivskyy@intel.com>

ozhuraki committed:
cmake: clang: Don't unconditionally use gcc's fortify

clang doesn't understand fortify at all, provide no op macro,
in order to handle the request to fortify in a generic way.

Signed-off-by: Oleg Zhurakivskyy <oleg.zhurakivskyy@intel.com>

ozhuraki committed:
cmake: clang: Provide toolchain_cc_nostdinc() macro

native_posix should build with standard includes.

Signed-off-by: Oleg Zhurakivskyy <oleg.zhurakivskyy@intel.com>

ozhuraki committed:
cmake: Set TOOLCHAIN_HOME for Zephyr's SDK

Host-tools don't unconditionally set TOOLCHAIN_HOME anymore,
but in case Zephyr's SDK toolchain is used, set TOOLCHAIN_HOME.

Signed-off-by: Oleg Zhurakivskyy <oleg.zhurakivskyy@intel.com>

ozhuraki commented:
@SebastianBoe @nashif 

Reworked all that we discussed, please take a look.

SebastianBoe commented:
Several comments are not addressed. Have a look at: https://github.com/zephyrproject-rtos/zephyr/pull/14077/files

ozhuraki commented:
@SebastianBoe 
> Several comments are not addressed. Have a look at: https://github.com/zephyrproject-rtos/zephyr/pull/14077/files
> 
> EDIT: Actually, just one.

OK. Sorry, it just slipped off my attention. Checking that.

ozhuraki commented:
@SebastianBoe @nashif

Reworked that remaining one, please take a look.

SebastianBoe commented on arch/posix/CMakeLists.txt:
```diff
@@ -1,12 +1,13 @@
   -include ${ZEPHYR_BASE}/arch/posix/include/posix_cheats.h
   )
 
+toolchain_cc_freestanding_options()
```
Intents need to be documented at the invocation.

Note that when I remarked that this should be a cc option, I actually meant that you could/should write

zephyr_cc_option(-fno-freestanding)

but this works as well.

SebastianBoe commented on cmake/compiler/gcc/target_freestanding.cmake:
```diff
@@ -0,0 +1,8 @@
+
+macro(toolchain_cc_freestanding_options)
+
+  if("${ARCH}" STREQUAL "posix")
```
This if-statement should be dropped, it is up to the user to determine for which ARCH's that freestanding should be turned off. This is already handled because arch/posix/CMakeLists.txt is only executed when ARCH STREQUAL "posix".

ozhuraki commented:
@SebastianBoe @nashif

OK. Documented the macro invocation and removed “if…posix” (actually a good catch), please take a look.

ozhuraki commented:
@SebastianBoe @nashif

ping

ozhuraki commented:
@SebastianBoe

Sebastian, if not too busy, could you please merge this one.

There’s a bit more in order to support libFuzzer for the native_posix, it would be nice if these would go through first.

SebastianBoe commented:
I don't have merge-rights unfortunately, but @nash does.

aescolar commented:
@ozhuraki We are in code freeze, so in principle only bugfixes and documentation improvements will be merged until the release.

ozhuraki commented:
@aescolar @SebastianBoe @nashif 

> We are in code freeze, so in principle only bugfixes and documentation improvements will be merged until the release.

OK. In principle, this can wait.

nashif commented:
@ozhuraki rebase please.

ozhuraki commented:
@nashif rebased, please take a look.

nashif commented:
@ozhuraki I am getting this on fedora 29:

```
[70/71] Linking C executable zephyr/zephyr.elf
/usr/bin/ld: skipping incompatible /usr/bin/../lib/gcc/x86_64-redhat-linux/8/libgcc.a when searching for -lgcc
/usr/bin/ld: skipping incompatible /usr/bin/../lib/gcc/x86_64-redhat-linux/8/libgcc.a when searching for -lgcc
/usr/bin/ld: skipping incompatible /usr/bin/../lib/gcc/x86_64-redhat-linux/8/libgcc_s.so when searching for -lgcc_s
/usr/bin/ld: skipping incompatible /usr/bin/../lib/gcc/x86_64-redhat-linux/8/libgcc.a when searching for -lgcc
/usr/bin/ld: skipping incompatible /usr/bin/../lib/gcc/x86_64-redhat-linux/8/libgcc_s.so when searching for -lgcc_s
```

nashif reviewed (changes requested):
also, please add llvm to supported toolchain in:

boards/posix/native_posix/native_posix.yaml

ozhuraki committed:
boards: native_posix: Add llvm to supported toolchains

Add llvm to supported toolchains.

Signed-off-by: Oleg Zhurakivskyy <oleg.zhurakivskyy@intel.com>

ozhuraki committed:
compiler: clang: Don't link libgcc for native_posix

libgcc isn't used by native_posix.

Signed-off-by: Oleg Zhurakivskyy <oleg.zhurakivskyy@intel.com>

ozhuraki commented:
@nashif 
> also, please add llvm to supported toolchain

Added.

> I am getting this on fedora 29

Resolved, please take a look.

nashif commented:
@ozhuraki looks better now, but if you rebase on top of latest master...

```
clang-7: error: unknown argument: '-fmacro-prefix-map=/home/nashif/Work/zephyrproject/zephyr=.'
```

ozhuraki commented:
@nashif 
> but if you rebase on top of latest master...

OK, I will take a look.

ozhuraki committed:
cmake: llvm: Invalidate toolchain capability cache

Invalidate toolchain capability cache on toolchain configuration.

Signed-off-by: Oleg Zhurakivskyy <oleg.zhurakivskyy@intel.com>

ozhuraki commented:
@nashif 

OK, reproduced. Looks like toolchain capability cache is in the way on Fedora. On Ubuntu, I don’t see such problem.

There’s probably less intrusive way to overpass this.

marc-hb commented:
> clang-7: error: unknown argument: '-fmacro-prefix-map=/home/nashif/Work/zephyrproject/zephyr=.'

@benpicco also hit this weird cache issue with just gcc, no clang involved. See his detailed and useful report in issue #15551

PS: I just discovered this PR in the git log. Considering my deterministic builds work in progress (see links in PR #14593) please copy me on major cmake and build PRs like this one, thanks!

SebastianBoe commented:
Hi, @ozhuraki , this patch:

https://github.com/zephyrproject-rtos/zephyr/commit/cbe74d46e042297a812330983cadbccc57c377c9

slipped past review.

Can you elaborate more on the rationale, or steps-to-reproduce the issue faced than this:

> OK, reproduced. Looks like toolchain capability cache is in the way on Fedora. On Ubuntu, I don’t see such problem.

ozhuraki commented:
@SebastianBoe 
> steps-to-reproduce the issue

Changing in between gcc and clang for native_posix on Fedora was reliably triggering this.

On Ubuntu it wasn’t reproducible.

Removing the ${USER_CACHE_DIR} was fixing the issue. So, possibly something is wrong/incompatible with the toolchain capability database in such cases, hence this patch.

marc-hb commented:
>  "possibly" something is wrong

There should at the very least be a comment in the source saying this was a blind shot and at what.
